### PR TITLE
Remove domain parameters

### LINKED ISSUE #6495: Remove support for domain parameters
Domain parameters are an experimental feature which is:

* Implemented only for RSA keys, in the following ways:
    * On key creation (import or generate), set the public exponent (without this interface, it has to be 65537).
    * In `psa_get_key_attributes`, get the public exponent. This is not useful since you can get it from the public key and you basically never need it separately.
* Specified for DH and DSA with finite fields, which are not implemented in Mbed TLS. (FFDH is coming, but without support for custom groups, so not using domain parameters yet).

At this point, it's unlikely that we'll ever implement custom groups for FFDH. We won't implement FFDSA. So we don't really have a planned use for domain parameters. They might come up again with other algorithms (e.g. for PQC, depending on exactly standardization goes), but it's not a given and if they do a different design might work better.

For RSA, the use of domain parameters was grafted on: we needed a way to select a custom public exponent when generating a key, the key attributes were the only way to pass information, and there was this field in the attributes that wasn't used for anything else and kind of fit. If we remove domain parameters, we need [a different way to specify the public exponent when generating an RSA key](https://github.com/Mbed-TLS/mbedtls/issues/6494).

Goal of this task: remove support for domain parameters from the API. Prerequisite: implement the alternative method for RSA generation (https://github.com/Mbed-TLS/mbedtls/pull/8815), and get to a point where we're comfortable with removing support for the old method. Once that's done, the domain parameter functions will be just stubs that do nothing.

Definition of done: domain parameters are removed from the public interface. If they are still present internally, file a follow-up task to clean up.

Advantages:

* Simplify the code.
* Will make https://github.com/Mbed-TLS/mbedtls/issues/4095 and https://github.com/Mbed-TLS/mbedtls/issues/3887 moot.
* Reduce the code size. See also https://github.com/Mbed-TLS/mbedtls/issues/6496

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Remove domain parameters from the PSA API, since we have an [alternative way of choosing a custom public exponent for RSA keys](https://github.com/Mbed-TLS/mbedtls/pull/8815) and we don't intend to use them for anything else. Resolve #6495.

Continues https://github.com/Mbed-TLS/mbedtls/pull/8815, will be rebased once that is merged.

Follow-up: https://github.com/Mbed-TLS/mbedtls/issues/6496 (started in https://github.com/Mbed-TLS/mbedtls/pull/7743, but removing domain parameters changing the deal).

## PR checklist

- [x] **changelog** provided
- [x] **backport** no
- [x] **tests** no

### CODE REVIEW

**Review by mpg — COMMENTED** [review]
Looks good to me except for a naming issue.

File: library/psa_crypto.c [hunk 1]
```diff
@@ -7576,11 +7426,8 @@ psa_status_t psa_generate_key_internal(
          * that mbedtls_psa_rsa_generate_key() gets e via a new
          * parameter instead. */
         psa_key_attributes_t override_attributes = *attributes;
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/psa_crypto.c:7428]: Oh, duh, yes. And also the comment above is obsolete, that was the whole point of this change.

**Review by valeriosetti — APPROVED** [review]
Thanks for addressing my comments. LGTM :)

File: library/psa_crypto.c [hunk 2]
```diff
@@ -1283,55 +1237,7 @@ psa_status_t psa_get_key_attributes(mbedtls_svc_key_id_t key,
...
-    }
-
-    if (status != PSA_SUCCESS) {
-        psa_reset_key_attributes(attributes);
-    }
-
-    unlock_status = psa_unregister_read_under_mutex(slot);
-
-    return (status == PSA_SUCCESS) ? unlock_status : status;
+    return psa_unregister_read(slot);
```
Comment (mpg) [inline library/psa_crypto.c:1240]: The code this replaces was calling `psa_unregister_read_under_mutex()`. Is the change intentional or was it missed during rebase?

**Review by mpg — COMMENTED** [review]
Looks good to me except for a possible rebase mistake. Please clarify If this was intentional.

File: library/psa_crypto.c (hunk 2 above)
Comment (gilles-peskine-arm) [reply library/psa_crypto.c]: Rebase mistake. Not caught by testing yet (thanks Ryan), would have caused headaches once the Tsan testing is in place. Thanks for catching it!

**Review by valeriosetti — APPROVED** [review]

**Review by mpg — APPROVED** [review]

**Pull request merged**